perf/perf_metric.py

```diff
@@ -0,0 +1,121 @@
+
+
+class perf_metric(Test):
+
+    """
+    This tests all metric/metric group events
+    :avocado: tags=perf,metric,events
+    """
+    # Initializing fail command list
+    fail_cmd = list()
+
+    def _create_all_metric_events(self, match):
+        cmd = "perf list %s" % match
+        output = process.system_output(cmd, shell=True, ignore_status=True)
+        for ln in output.decode().split('\n'):
+            ln = ln.strip()
+            # Removing empty line, header and comment
+            if not ln or "List of pre-defined events" in ln or "[" in ln or\
+               "Metrics:" in ln or "Metric Groups:" in ln:
+                continue
+            else:
+                self.list_of_metric_events.append(ln)
+
+    def setUp(self):
+        """
+        Setup checks :
+        0. Processor should be ppc64.
+        1. Install perf package
+        2. Check for metric/metric group
+        3. When found metric/metric group run all the events
+        """
+        smm = SoftwareManager()
+
+        detected_distro = distro.detect()
+        if 'ppc64' not in detected_distro.arch:
+            self.cancel("Processor is not PowerPC")
+
+        deps = ['gcc', 'make']
+        if 'Ubuntu' in detected_distro.name:
+            deps.extend(['linux-tools-common', 'linux-tools-%s'
+                         % platform.uname()[2]])
+        elif detected_distro.name in ['rhel', 'SuSE', 'fedora', 'centos']:
+            deps.extend(['perf', 'numactl'])
+        else:
+            self.cancel("Install the package for perf supported by %s"
+                        % detected_distro.name)
+        for package in deps:
+            if not smm.check_installed(package) and not smm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+
+        # Collect all metric events and metric group events
+        self.list_of_metric_events = []
+        self._create_all_metric_events('metric')
+        self._create_all_metric_events('metricgroup')
+        if not self.list_of_metric_events:
+            self.cancel("perf tool Metric events not found.")
+
+        # Clear the dmesg to capture the delta at the end of the test.
+        process.run("dmesg -c", sudo=True)
+
+    def _run_cmd(self, option):
+        for line in self.list_of_metric_events:
+            cmd = "perf stat %s %s sleep 1" % (option, line)
+            rc, output = process.getstatusoutput(cmd, ignore_status=True,
+                                                 shell=True, verbose=True,
+                                                 allow_output_check='combined')
+            # When the command failed, checking for expected failure or not.
+            if rc:
+                found_imc = False
+                found_hv_24_7 = False
+                for ln in output.split('\n'):
+                    if "hv_24x7" in ln:
+                        found_hv_24_7 = True
+                        break
```

Comment:
Are `hv_24x7` and `imc` mutually exclusive? Or will the string `hv_24x7` always comes after `imc` in the loop?  If either is not the case, then the break needs to be removed here since the loop will be exited if `hv_24x7` is found first.

Reply:
> Are hv_24x7 and imc mutually exclusive?
Either of the events will be there in the error log. So coming out of the loop after seeing one event.